### ISSUE TITLE
Matching patches to residues

### ISSUE THREAD

**Issue by peastman** [ParmEd/ParmEd#1276]
I'm investigating a problem (https://github.com/openmm/openmm/issues/3815) in which the ACE patch for the CHARMM polar force field didn't get converted correctly.  Part of the problem is in how it matches patches to residues.

In an OpenMM force field, a patch definition needs to include information that isn't specified in the CHARMM parameter files.  For example, when it lists an `ATOM` record, is that a new atom to be added or an existing atom to be modified?  When it lists a `BOND`, is that a new bond to be added, or is it replacing what was previously an external bond to a different residue?  To determine that information, we apply the patch to an actual residue and compare it to the original unpatched residue.

But which residue to choose?  CHARMM parameter files also don't specify which residues each patch is meant to be applied to.  So we try every possible residue.  Different residues may lead to different definitions for the patch, so we take a vote.  Whatever patch definition the largest number of residues produce, that's what we use.

For the ACE residue, that isn't working.  It's meant to be applied to amino acids, but in principle it could be applied to any residue containing an atom called `N`, which is a lot of them.  For example, lots of types of lipids.  In fact, there turn out to be more of those than there are amino acids.  And since `N` doesn't form an external bond in lipids, it therefore omits the `<RemoveExternalBond>` tag which is essential for applying it to amino acids.

One possible solution to this problem is that when it applies the patch to residues for the purpose of defining the patch information, it only considers ones that came from the same file.  The ACE patch is defined in toppar_drude_master_protein_2019h.str, so it would only consider residues found in that same file, not ones in other files like toppar_drude_lipid_2017c.str.

Do you think that would work?  Are there any situations it would break, where a single file contains patches but no residues those patches are meant to be applied to?  Can you think of a better solution?

**Comment by peastman** [comment]
I don't see any existing mechanism to tell what file a patch or residue was loaded from, so presumably that would require creating one.

**Comment by peastman** [comment]
@swails do you have any suggestions on the best way to handle this?  We need to come up with something, since it's a blocker for OpenMM 8.0.  If absolutely necessary I can put in a hack just for that one force field, but of course I'd much rather fix it in a clean, general way.

**Comment by swails** [comment]
The solution you describe seems a little brittle to me, since I imagine (although don't know) that it's possible for people to create patches in str files that should be applied to residues outside that file.  We can certainly add a 

I've looked for a copy of the original CHARMM files but couldn't find them -- could you point me in that direction?  Does the file not indicate that it replaces an external bond from the N atom?

**Comment by peastman** [comment]
You can get the files from http://mackerell.umaryland.edu/charmm_ff.shtml.  The links on the main CHARMM website seem to be broken.

Here's the definition of the ACE patch.

```
PRES ACE           0.000 ! acetylated N-terminus in di- and polypeptides 
                         ! use in generate statement 
! 
! HY1 HY2 HY3 
!    \ | / 
!     CAY 
!      | 
!      CY=OY 
!      | 
! 
GROUP 
 
ATOM CAY  CD33C   -0.102  ALPHA -1.894  THOLE 2.122 
ATOM HY1  HDA3A    0.048 
ATOM HY2  HDA3A    0.048 
ATOM HY3  HDA3A    0.048 
ATOM CY   CD2O1A   0.497  ALPHA -0.675  THOLE 0.295 
ATOM OY   OD2C1A   0.000  ALPHA -0.651  THOLE 0.310 
ATOM LPY1 LPDO1   -0.312 
ATOM LPY2 LPDO1   -0.227 
 
BOND N      CY     CAY    CY     CY     OY 
BOND CAY    HY1    CAY    HY2    CAY    HY3 
BOND OY     LPY1   OY    LPY2 
 
IMPR CY     CAY    N      OY     N      CY     CA     HN 
CMAP CY     N      CA     C      N      CA     C      +N 
CMAP CY     N      CA     C      N      CA     C      NT !for dipeptide 
 
LONEPAIR relative LPY1 OY CY CAY distance 0.30 angle 91.0 dihe   0.0 
LONEPAIR relative LPY2 OY CY CAY distance 0.30 angle 91.0 dihe 180.0 
ANISOTROPY OY CY LPY1 LPY2 A11  0.82322 A22  1.14332 
 
ACCEPTOR OY CY 
IC CY   N    CA   C     0.0000  0.0000  -60.0000  0.0000  0.0000 
IC CY   CA   *N   HN    0.0000  0.0000  180.0000  0.0000  0.0000 
IC CAY  CY   N    CA    0.0000  0.0000  180.0000  0.0000  0.0000 
IC N    CAY  *CY  OY    0.0000  0.0000  180.0000  0.0000  0.0000 
IC OY   CY   CAY  HY1   0.0000  0.0000  180.0000  0.0000  0.0000 
IC OY   CY   CAY  HY2   0.0000  0.0000   60.0000  0.0000  0.0000 
IC OY   CY   CAY  HY3   0.0000  0.0000  -60.0000  0.0000  0.0000 
 ```

**Comment by swails** [comment]
Yuck.  OK.  I would look at this and see that there are bonds defined between the atoms in that patch and `N`, which is not in the patch.  I would take this to mean that the patch can be applied to any residue that has an atom named `N` that forms an external bond with the adjacent residue.

This should exclude lipids and RNA/DNA residues and limit the residue combinations you need to try the same way as if you'd restricted to just residues in the same file.

I don't think I know as much about CHARMM patches as you do, though -- would this cause problems with other patches?  If I recall correctly, patches that would "force" a bond between two residues using atoms that don't normally bond externally would typically have some other indicator that it was doing that (the disulfide patch comes to mind).

I think it would be helpful to understand how CHARMM itself applies patches so we can emulate its behavior.

**Comment by peastman** [comment]
>  I would take this to mean that the patch can be applied to any residue that has an atom named N that forms an external bond with the adjacent residue.

I don't think we can rely on the part about the external bond.  Patches can be used to make internal changes to a single residue.  For example, this one modifies the side chain of a ASP residue.

```
PRES ASPP          0.000 ! patch for protonated aspartic acid, proton on od2 
                         ! via acetic acid, use in a patch statementand 
                         ! follow with AUTOgenerate ANGLes DIHEdrals command 
! 
! HB1    OD1 
!  |    // 
! -CB--CG 
!  |     \ 
! HB2     OD2-HD2 
! 
```

Possibly we could say that patches are *usually* applied to terminal residues, and therefore we should prefer a version that removes an external bond to one that doesn't.  It's a hack, but anything we do is likely to be a hack to some extent.

> I think it would be helpful to understand how CHARMM itself applies patches so we can emulate its behavior.

As I understand it, CHARMM never applies patches automatically.  You manually execute the [PATCH](https://charmm-gui.org/charmmdoc/struct.html#patch-command-to-modify-psf) command to tell it to modify the topology in a particular way.  So it never needs to figure out where patches are or aren't appropriate.  It's up to the user to tell it.

**Comment by swails** [comment]
Oh that's really frustrating.

I think trying to find a general solution is a fool's errand here.  Can we configure John's scripts that do the conversion to add a delete-external-bond directive to that patch?

It's definitely not satisfying, but would work and is actually closer to the CHARMM ethos of "specify it yourself"

**Comment by peastman** [comment]
I agree that finding a robust, general purpose solution probably isn't possible.  Fortunately, we don't really need that.  If we can just find a hack that isn't too horrible and that produces correct results for the force fields we care about (which at the moment only means CHARMM36 and CHARMM-Drude), that should be good enough.  I believe that either basing it on source file, or preferring versions that remove external bonds, would do that.

> Can we configure John's scripts that do the conversion to add a delete-external-bond directive to that patch?

If I understand you, that would require editing the force field after ParmEd is done writing it out?  That's going to be a lot more complicated than building the logic into ParmEd.

**Comment by swails** [comment]
> If I understand you, that would require editing the force field after ParmEd is done writing it out?

Not really.  What I had in mind was that John's conversion utility (which uses ParmEd) would read the CHARMM files, make changes to the ParmEd residue/patch objects defined by an overrides in the config file for that force field, then write them out to an OpenMM XML file.

This may be an outdated mental model of how this operates though -- where is the process by which these files are created?

**Comment by peastman** [comment]
It does some very minimal editing, just setting a single attribute of each residue and patch.  This involves changes at a deeper level, to the logic in `OpenMMParameterSet._write_omm_patches()`.  That's where it tries to apply each patch to every possible residue, makes a list of the patching instructions that result from each one, and selects the set of instructions that results from the largest number of residues.  We need to change the logic so it will pick the set of instructions to use differently (either favoring ones that came from the same file, or favoring ones that involve changes to external bonds, or something else).

**Comment by swails** [comment]
Can we break it up so that we create protein, nucleic acid, and lipid parameters separately (into separate XML files)?

That way the ACE/NME patches would apply to every residue in its set, but that set would only be proteins (and so would do the right thing).

**Comment by peastman** [comment]
I tried processing each of the files separately.  When I process the lipid file I get an exception:

> RuntimeError: No improper found for improper ('C29', 'C28', 'C210', 'H9R') in residue DOPC (types were [['CD2C1A'], ['CD32A'], ['CD2C1A'], ['HD2C1A']])

I assume this is because the files aren't meant to stand alone, with the lipid file depending on definitions found in the protein file?  Is there a way to make that work?

**Comment by swails** [comment]
What we can probably do is read each file, then update *just* the parameter definitions from the other files so they can always be found (without updating the residue templates).

Does that sound reasonable?

**Comment by peastman** [comment]
I'm not sure I understand, but if that will have the right effect, I'll take your word for it!

**Comment by peastman** [comment]
Can you describe what I need to change?  I don't completely understand.

**Comment by swails** [comment]
You can do something like this:

```python
# Now overwrite the residues to just the ones contained in each file
for str_file in [protein_str, nucleic_str, lipid_str]:
    # Load all of the parameters into the same set
    all_parameters = CharmmParameterSet(protein_str, nucleic_str, lipid_str, ...)

    all_parameters.residues = CharmmParameterSet(str_file).residues
    # Convert to OpenMM parameter sets -- I forget the exact incantation here
```

In this way, all of the parameters will be present even if they're in a different file, but the patches will only apply to the residues in each file.

Would that be an acceptable workaround?

**Comment by peastman** [comment]
Wouldn't that lead to the same force definitions being repeated in each file?  That would produce incorrect results when you loaded all three files together and used them to parametrize a system.

For what it's worth, I tried the change to make it prefer patches that remove external bonds.  It involved only very minor changes to `_write_omm_patches()` and seems to work.

**Comment by peastman** [comment]
For standard forces, it merges duplicate definitions from different files.  But for custom forces, every tag is treated independently and produces a separate Force object.  CHARMM uses custom forces for some terms.  It implements impropers with

```xml
  <CustomTorsionForce energy="k*(theta-theta0)^2">
```

You'll get three of those tags, one in each file, and every one will contain the complete set of definitions to match every improper in every type of molecule.  You'll end up with three CustomTorsionForces, and an effective force constant that's three times too high.
